### PR TITLE
Tools to generate geographic coverage area data from ISO 3166-1/2 codes

### PR DESCRIPTION
**Pull request by vkrause**
Gives us better quality and easier to maintain polygons over hand-drawn ones (or polygons at all for those we don't have any for at the moment), in cases we have no better geographic coverage area data.

Not applied to all files yet, using DB for demonstration for now.

### CODE REVIEW

File: tools/pretty-json.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+parser.add_argument('filename')
+arguments = parser.parse_args()
+
+# read as JSON and normalize to standard formatting
+with open(arguments.filename, 'r') as f:
+    j = json.loads(f.read())
+s = json.dumps(j, indent = 2)
+
+# fold arrays of scalar values into one line
+s = re.sub(r'\[\n +(\"[A-Za-z-]+\"|[\d\.-]+)', r'[\1', s)
```
Comment (derhuerst) [inline tools/pretty-json.py]: What do you think about doing this transformation only inside `coverage.area`?

**Review by derhuerst — COMMENTED** [review]
I'm not familiar enough with Python to judge off the top of my head if all `import`s are built-ins. If there are some packages from PyPi being used, shall we add a `requirements.txt` file, so that we always get compatible versions?

**Comment by vkrause** [comment]
> I'm not familiar enough with Python to judge off the top of my head if all `import`s are built-ins. If there are some packages from PyPi being used, shall we add a `requirements.txt` file, so that we always get compatible versions?

Right, I tried to limit this to basic/built-in stuff so this works on any standard Python installation without needing extra dependencies. Makes some of the geometry simplification math a bit more manual though.

File: tools/pretty-json.py (hunk 1 above)
Comment (vkrause) [reply tools/pretty-json.py]: We could do that, but that would probably mean writing a full custom JSON serializer, simple post-processing on the JSON string isn't going to be enough for that I suspect.

The difference are the usually very short arrays for Hafas product bitmasks or supported languages/coverage regions for example, splitting those per line doesn't add that much readability/diff-ability.

So unsure whether that's worth the effort?

Comment (derhuerst) [reply tools/pretty-json.py]: probably not...

**Comment by derhuerst** [comment]
Shall we add a `build.sh` script that calls the scripts in the right order? So that others don't need to wrap their head around how they work.

**Comment by vkrause** [comment]
> Shall we add a `build.sh` script that calls the scripts in the right order? So that others don't need to wrap their head around how they work.

Added a build script, and additional documentation.

File: tools/build.sh [hunk 2]
```diff
@@ -0,0 +1,27 @@
+#!/bin/sh
+# SPDX-FileCopyrightText: 2023 Volker Krause <vkrause@kde.org>
+# SPDX-License-Identifier: MIT
+
+scriptdir=`dirname "${BASH_SOURCE[0]:-$0}"`
```
Comment (derhuerst) [inline tools/build.sh]: `$BASH_SOURCE` isn't defined in an Alpine Linux `/bin/sh`, only in `/bin/bash`.

File: tools/build.sh [hunk 3]
```diff
@@ -0,0 +1,27 @@
+#!/bin/sh
+# SPDX-FileCopyrightText: 2023 Volker Krause <vkrause@kde.org>
+# SPDX-License-Identifier: MIT
+
```
Comment (derhuerst) [inline tools/build.sh]:
```suggestion

set -e -u -o pipefail

```

File: tools/build.sh [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+scriptdir=`dirname "${BASH_SOURCE[0]:-$0}"`
+datadir=$scriptdir/../data
+
+
+
+for j in `find $datadir -name "*.json"`; do
+    country=`echo "$j" | sed -e 's,^.*/\([a-z][a-z]\)/.*$,\1,'`
```
Comment (derhuerst) [inline tools/build.sh]: This seems very brittle.

**Review by derhuerst — CHANGES_REQUESTED** [review]
There are some problems with `build.sh`. What do you think about using Python for it, too?

File: tools/build.sh (hunk 2 above)
Comment (vkrause) [reply tools/build.sh]: Right, that's why there is the fallback to `$0`, which should be available everywhere (but might not contain the right thing if you get creative in calling the script).

File: tools/build.sh (hunk 4 above)
Comment (vkrause) [reply tools/build.sh]: It is. The alternative I was considering is an explicit mapping to the bounding box for each file that needs one, much more flexible but also more maintenance/duplication. Not pretty either.

**Comment by vkrause** [comment]
> There are some problems with `build.sh`. What do you think about using Python for it, too?

That would likely be the more robust and easier choice, and Python is assumed to be available for all this anyway.

**Comment by 1Maxnet1** [comment]
> > I'm not familiar enough with Python to judge off the top of my head if all `import`s are built-ins. If there are some packages from PyPi being used, shall we add a `requirements.txt` file, so that we always get compatible versions?
> 
> Right, I tried to limit this to basic/built-in stuff so this works on any standard Python installation without needing extra dependencies. Makes some of the geometry simplification math a bit more manual though.

At least for me I needed to install requests library, as long as it does stay with this little dependencies I wouldn't go for a requirements.txt either, but if we choose to have more of them, would add it anyway.

**Comment by 1Maxnet1** [comment]
Another thing is that currently the python scripts are not following the widespread PEP8 Guidelines, which I would propose to do (can offer a PR to this work branch if you wish to).

**Comment by vkrause** [comment]
> Another thing is that currently the python scripts are not following the widespread PEP8 Guidelines, which I would propose to do (can offer a PR to this work branch if you wish to).

That is most certainly not intentional and only due to my very limited Python experience :)

**Comment by 1Maxnet1** [comment]
> > Another thing is that currently the python scripts are not following the widespread PEP8 Guidelines, which I would propose to do (can offer a PR to this work branch if you wish to).
> 
> That is most certainly not intentional and only due to my very limited Python experience :)

Okay, so if you want to apply some rules I created #70 :)

**Comment by vkrause** [comment]
> There are some problems with `build.sh`. What do you think about using Python for it, too?

Sorry for the long delay, finally got around to do that now.

**Review by derhuerst — APPROVED** [review]
Not a Python export. I'm fine with the general idea behind this PR!

**Pull request merged**